activation/nipost.go

```diff
@@ -237,35 +242,32 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.NIPos
 	// Phase 1: query PoET services for proofs
 	if nb.state.PoetProofRef == types.EmptyPoetProofRef {
 		now := time.Now()
-		// Deadline: the end of the publish epoch (with a safety margin of `GracePeriod`). If we do not publish within
-		// the publish epoch we won't receive any rewards in the target epoch.
-		if publishEpochEnd.Before(now) {
-			return nil, 0, fmt.Errorf("%w: deadline to query poet proof for pub epoch %d exceeded (deadline: %s, now: %s)", ErrATXChallengeExpired, challenge.PublishEpoch, publishEpochEnd, now)
+		// Deadline: the end of the publish epoch minus the cycle gap. A node that is setup correctly (i.e. can
+		// generate a PoST proof within the cycle gap) has enough time left to generate a post proof and publish.
+		if poetProofDeadline.Before(now) {
+			return nil, fmt.Errorf("%w: deadline to query poet proof for pub epoch %d exceeded (deadline: %s, now: %s)", ErrATXChallengeExpired, challenge.PublishEpoch, poetProofDeadline, now)
 		}
-		getProofsCtx, cancel := context.WithDeadline(ctx, publishEpochEnd)
-		defer cancel()
 
 		events.EmitPoetWaitProof(challenge.PublishEpoch, challenge.TargetEpoch(), time.Until(poetRoundEnd))
-		poetProofRef, membership, err := nb.getBestProof(getProofsCtx, nb.state.Challenge)
+		poetProofRef, membership, err := nb.getBestProof(ctx, nb.state.Challenge)
 		if err != nil {
-			return nil, 0, &PoetSvcUnstableError{msg: "getBestProof failed", source: err}
+			return nil, &PoetSvcUnstableError{msg: "getBestProof failed", source: err}
 		}
 		if poetProofRef == types.EmptyPoetProofRef {
-			return nil, 0, &PoetSvcUnstableError{source: ErrPoetProofNotReceived}
+			return nil, &PoetSvcUnstableError{source: ErrPoetProofNotReceived}
 		}
 		nb.state.PoetProofRef = poetProofRef
 		nb.state.NIPost.Membership = *membership
 		nb.persistState()
 	}
 
 	// Phase 2: Post execution.
-	var postGenDuration time.Duration = 0
 	if nb.state.NIPost.Post == nil {
 		now := time.Now()
-		// Deadline: the end of the publish epoch (with a safety margin of `GracePeriod`). If we do not publish within
+		// Deadline: the end of the publish epoch. If we do not publish within
 		// the publish epoch we won't receive any rewards in the target epoch.
 		if publishEpochEnd.Before(now) {
-			return nil, 0, fmt.Errorf("%w: deadline to publish ATX for pub epoch %d exceeded (deadline: %s, now: %s)", ErrATXChallengeExpired, challenge.PublishEpoch, publishEpochEnd, now)
+			return nil, fmt.Errorf("%w: deadline to publish ATX for pub epoch %d exceeded (deadline: %s, now: %s)", ErrATXChallengeExpired, challenge.PublishEpoch, publishEpochEnd, now)
 		}
 		postCtx, cancel := context.WithDeadline(ctx, publishEpochEnd)
```

Comment:
IMHO, this check shouldn't be necessary. Isn't it equivalent to checking if `ctx` is canceled already? `AtxBuilder` knows the deadline to publish an ATX and it should time out the passed ctx properly, what do you think?

Reply:
As discussed offline, we decided to keep this deadline for now and possibly revisit it at a later time (e.g. when canceling `GenerateProof` becomes possible).